### PR TITLE
Avoid Stream API in RowBlockBuilder constructor

### PR DESCRIPTION
**Pull request by wenleix**
Row/JSON to row cast will constructor a new RowBlockBuilder
    for each invocation.

### CODE REVIEW

**Comment by wenleix** [comment]
Benchmark shows about 20% improvements with current ROW to ROW cast:

Before:

```
Benchmark                        Mode  Cnt    Score   Error  Units
BenchmarkRowToRowCast.benchmark  avgt   60  302.564 ± 2.344  ns/op
```

After:

```
Benchmark                        Mode  Cnt    Score   Error  Units
BenchmarkRowToRowCast.benchmark  avgt   60  252.795 ± 1.075  ns/op
```

An alternative plan would be implement direct output builder for ROW to ROW cast, based on @gerlou's work. 

Thoughts? @haozhun

**Comment by findepi** [comment]
> Avoid Stream API in RowBlockBuilder constructor 

as a rule of thumb, we generally avoid streams in Block, Page classes, and in Operators (except perhaps during construction and cleanup)

File: presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java [hunk 1]
```diff
@@ -45,15 +45,14 @@
...
     private boolean currentEntryOpened;
 
-    public RowBlockBuilder(List<Type> fieldTypes, BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    public static RowBlockBuilder createRowBlockBuilder(List<Type> fieldTypes, BlockBuilderStatus blockBuilderStatus, int expectedEntries)
     {
-        this(
-                blockBuilderStatus,
-                fieldTypes.stream()
-                        .map(type -> type.createBlockBuilder(blockBuilderStatus, expectedEntries))
-                        .toArray(BlockBuilder[]::new),
```
Comment (findepi) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java]: You could have retained the constructor and covert the Stream-based code into plain static method

**Comment by dain** [comment]
@wenleix, which JVM did you benchmark on?  Can you run the benchmark in 10?... just curious

**Comment by wenleix** [comment]
@dain : It's running on JDK 8. Let's try JDK 10 :)

**Comment by dain** [comment]
You got a style failure in Travis:
```
src/test/java/com/facebook/presto/operator/scalar/BenchmarkRowToRowCast.java:[107,72] (blocks) LeftCurly: '{' at column 72 should be on a new line.
```

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkRowToRowCast.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkRowToRowCast
+{
+    private static final int POSITION_COUNT = 100_000;
+
+    @Benchmark
+    @OperationsPerInvocation(POSITION_COUNT)
+    public List<Optional<Page>> benchmark(BenchmarkData data)
+    {
+        return ImmutableList.copyOf(data.getPageProcessor().process(SESSION, new DriverYieldSignal(), data.getPage()));
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkRowToRowCast.java]: @nezihyigitbasi  : Copied from other benchmarks, such as `BenchmarkArrayJoin`.

@haozhun  : do you have an idea why we do this for benchmarks in Presto ?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkRowToRowCast.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            return pageProcessor;
+        }
+
+        public Page getPage()
+        {
+            return page;
+        }
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkRowToRowCast.java]: @nezihyigitbasi : Seems to be  a Presto benchmark convention. My understanding is it guards benchmark to be run correctly (e.g. later we change some internal implementations)

**Comment by haozhun** [comment]
> Streams are known to have a performance hit, and we should avoid it in performance critical code. 

@nezihyigitbasi My take is to avoid stream entirely in the code base and ban it because
1. it's too hard to be able to consistently correctly identify the performance sensitivity of a piece of code;
2. the benefit `stream` provides is pretty minimal. 

I think my opinion on this topic is minority.

**Comment by wenleix** [comment]
Benchmark result on JDK 10 (MacBook Pro (15-inch, 2017)) :

Before

```
Benchmark                        Mode  Cnt    Score   Error  Units
BenchmarkRowToRowCast.benchmark  avgt   60  315.853 ± 6.785  ns/op
```

After

```
Benchmark                        Mode  Cnt    Score   Error  Units
BenchmarkRowToRowCast.benchmark  avgt   60  237.497 ± 2.824  ns/op
```

The performance is similar with Stream API while the non-stream implementation get slight boost.

**Pull request merged**